### PR TITLE
Adding a DFE Jupyter notebook

### PR DESCRIPTION
**Pull request by tonybruguier-google**
As requested, I added a Jupyter notebook that showcases the DFE code.

Note that the present version relies on under-review PR:
https://github.com/quantumlib/Cirq/pull/2918

I've never gone through a review for a notebook, so I am attaching a PDF of what it currently looks like.
[direct_fidelity_estimation.pdf](https://github.com/quantumlib/Cirq/files/4517980/direct_fidelity_estimation.pdf)

### CODE REVIEW

**Comment by tonybruguier-google** [comment]
Friendly ping on this PR.

I think @mpharrigan has the most context, but feedback from anyone is, of course, welcome.

**Comment by tonybruguier-google** [comment]
Hi Matt,

Do you have any comment on this PR? It's only a Jupyter notebook.

File: examples/direct_fidelity_estimation.ipynb [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+      "      │       │\n",
+      "2: ───X───H───@───\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Import Cirq, DFE, and create a circuit\n",
+    "\n",
+    "import cirq\n",
+    "import direct_fidelity_estimation as dfe\n",
```
Comment (mpharrigan) [inline examples/direct_fidelity_estimation.ipynb:38]: This import works when the notebook is run from the same directory as the example `direct_fidelity_estimation.py` in the examples directory; but won't work if this notebook is copied elsewhere.

In general, in Python I think these sorts of imports are discouraged in favor of using a proper package structure. Fixing this is well outside of the scope of the current PR. I have a feeling this project is eclipsing the bounds of a a simple "example". We can chat about various strategies to flesh this out better

File: examples/direct_fidelity_estimation.ipynb [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    "\n",
+    "qubits = cirq.LineQubit.range(3)\n",
+    "circuit = cirq.Circuit(cirq.CNOT(qubits[0], qubits[2]),\n",
+    "                       cirq.Z(qubits[0]),\n",
+    "                       cirq.H(qubits[2]),\n",
+    "                       cirq.CNOT(qubits[2], qubits[1]))\n",
+    "\n",
+    "print(circuit)"
```
Comment (mpharrigan) [inline examples/direct_fidelity_estimation.ipynb:46]: Consider using

```python
from cirq.contrib.svg import SVGCircuit
SVGCircuit(circuit)
```
for a nicer looking output

File: examples/direct_fidelity_estimation.ipynb [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    }
+   ],
+   "source": [
+    "# We run the DFE:\n",
+    "estimated_fidelity, intermediate_results = dfe.direct_fidelity_estimation(\n",
+    "        circuit,\n",
+    "        qubits,\n",
+    "        sampler,\n",
+    "        n_measured_operators=None,\n",
+    "        samples_per_term=0)\n",
```
Comment (mpharrigan) [inline examples/direct_fidelity_estimation.ipynb:80]: You should probably indicate what the sentinel values of `None` and `0` mean in this instance (the other arguments are self-explanatory)

edit: I see later you dive into this deeper. Maybe a short comment / text saying that you'll describe these parameters later in the notebook

File: examples/direct_fidelity_estimation.ipynb [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# What is happening under the hood?\n",
+    "Now, let's look at the `intermediate_results` and correlate what is happening in the code with the papers. The definition of fidelity is:\n",
+    "$$\n",
+    "F = F(\\hat{\\rho},\\hat{\\sigma}) = tr \\left(\\hat{\\rho} \\hat{\\sigma}\\right)\n",
```
Comment (mpharrigan) [inline examples/direct_fidelity_estimation.ipynb:92]: nit: `tr` -> `\mathrm{Tr}`

File: examples/direct_fidelity_estimation.ipynb [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    "$$\n",
+    "where $\\hat{\\rho}$ is the theoretical pure state and $\\hat{\\sigma}$ is the actual state. The idea of DFE is to write fidelity as:\n",
+    "$$F= \\sum _i \\frac{\\rho _i \\sigma _i}{d}$$\n",
+    "\n",
+    "where $d=4^{\\mathit{number-of-qubits}}$, $\\rho _i = tr \\left( \\hat{\\rho} P_i \\right)$, and $\\sigma _i = tr \\left(\\hat{\\sigma} P_i \\right)$. Each of the $P_i$ is a Pauli operator. We can then finally rewrite the fidelity as:\n",
```
Comment (mpharrigan) [inline examples/direct_fidelity_estimation.ipynb:97]: Tr

File: examples/direct_fidelity_estimation.ipynb [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+    "\n",
+    "$$F= \\sum _i Pr(i) . \\frac{\\sigma _i}{\\rho_i}$$\n",
```
Comment (mpharrigan) [inline examples/direct_fidelity_estimation.ipynb:99]: I don't think `.` is what you want here? `\cdot`? or just nothing and have the multiplication be implied

File: examples/direct_fidelity_estimation.ipynb [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    "Recall that the formula for DFE is:\n",
+    "$$F= \\sum _i Pr(i) . \\frac{\\sigma _i}{\\rho_i}$$\n",
+    "\n",
+    "But for Clifford circuits, we have $Pr(i) = \\frac{1}{d}$ and $\\rho_i = 1$ and thus the formula becomes:\n",
+    "$$F= \\frac{1}{d} \\sum _i \\sigma _i$$\n",
+    "\n",
+    "If we estimate by randomly sampling $N$ values for the indicies $i$ for $\\sigma_i$ we get:\n",
+    "$$\\hat{F} = \\frac{1}{N} \\sum_{j=1}^N \\sigma _{i(j)}$$\n",
+    "\n",
+    "Using the Bhatia–Davis inequality ([A Better Bound on the Variance, Rajendra Bhatia and Chandler Davis](https://www.jstor.org/stable/2589180)) and the fact that $0 \\le \\sigma_i \\le 1$, we have the variance of :\n",
```
Comment (mpharrigan) [inline examples/direct_fidelity_estimation.ipynb:195]: extreme nit: extraneous space before the colon ` :`

File: examples/direct_fidelity_estimation.ipynb [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    "\n",
+    "The code is optimized for Clifford circuits using tableaux\n",
+    "* Improved Simulation of Stabilizer Circuits (https://arxiv.org/pdf/quant-ph/0406196.pdf)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
```
[the remark this answers is not in the document]
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:21]: Done.

File: examples/direct_fidelity_estimation.ipynb [hunk 9]
```diff
@@ -0,0 +1,228 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Running the Direct Fidelity Estimation (DFE) algorithm\n",
+    "This example walks through the steps of running the direct fidelity estimation (DFE) algorithm as described in these two paper:\n",
```
[the remark this answers is not in the document]
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:8]: Done.

File: examples/direct_fidelity_estimation.ipynb (hunk 1 above)
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:38]: Added "examples." in front of the import

File: examples/direct_fidelity_estimation.ipynb (hunk 2 above)
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:46]: Done

File: examples/direct_fidelity_estimation.ipynb (hunk 4 above)
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:92]: Done

File: examples/direct_fidelity_estimation.ipynb (hunk 6 above)
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:99]: Done

Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:99]: Done

File: examples/direct_fidelity_estimation.ipynb (hunk 7 above)
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:195]: Done.

File: examples/direct_fidelity_estimation.ipynb (hunk 3 above)
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:80]: Done.

File: examples/direct_fidelity_estimation.ipynb [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    "\n",
+    "* Direct Fidelity Estimation from Few Pauli Measurements (https://arxiv.org/abs/1104.4695)\n",
+    "* Practical characterization of quantum devices without tomography (https://arxiv.org/abs/1104.3835)\n",
+    "\n",
+    "The code is optimized for Clifford circuits using tableaux\n",
```
[the remark this answers is not in the document]
Comment (tonybruguier-google) [reply examples/direct_fidelity_estimation.ipynb:13]: Done.

**Comment by balopat** [comment]
@googlebot please rescan

**Pull request merged**